File: src/main/java/LeetCode/JavaScript/Medium/SpiralMatrix.js
var spiralOrder = function(matrix) {
    const spiralArray = [];

    let horizontalMoves = matrix[0].length;
    let verticalMoves = matrix.length-1;
    let totalMoves = matrix.length*matrix[0].length;

    let xCoord = 0;
    let yCoord = 0;

    const directionArray = ["right", "down", "left", "up"];
    let directionIndex = 0;

    while(totalMoves) {
        let direction = directionArray[directionIndex];

        if (direction === "right" || direction === "left") {
            for (let i=0; i<horizontalMoves; i++) {
                spiralArray.push(matrix[yCoord][xCoord]);
                (direction === "right") ? xCoord++ : xCoord--;
            }
            totalMoves -= horizontalMoves;
            horizontalMoves--;
        } else {
            for (let i=0; i<verticalMoves; i++) {
                spiralArray.push(matrix[yCoord][xCoord]);
                (direction === "down") ? yCoord++ : yCoord--;
            }
            totalMoves -= verticalMoves;
            verticalMoves--;
        }

        (direction === "right" || direction === "up") ? yCoord++ : yCoord--;
        (direction === "right" || direction === "down") ? xCoord-- : xCoord++;
        directionIndex = (directionIndex===3) ? 0 : directionIndex+1;
    }

    return spiralArray;
};

File: src/main/java/LeetCode/Medium/SpiralMatrix.java
package LeetCode.Medium;

import java.util.ArrayList;
import java.util.List;

public class SpiralMatrix {
    public List<Integer> spiralOrder(int[][] matrix) {
        int xCoordinate = 0;
        int yCoordinate = 0;
        int totalIterations = matrix.length*matrix[0].length;

        int horizontalMaxMoves = matrix[0].length;
        int verticalMaxMoves = matrix.length-1;

        String[] directions = {"right", "down", "left", "up"};
        int directionIndex = 0;

        List<Integer> finalList = new ArrayList<>();

        while(totalIterations > 0) {
            if (directions[directionIndex].equals("right") || directions[directionIndex].equals("left")) {
                for(int i=0; i<horizontalMaxMoves; i++) {
                    finalList.add(matrix[yCoordinate][xCoordinate]);
                    xCoordinate = (directions[directionIndex].equals("right")) ? xCoordinate+1 : xCoordinate-1;
                    totalIterations--;
                }
                xCoordinate = (directions[directionIndex].equals("right")) ? xCoordinate-1 : xCoordinate+1;
                horizontalMaxMoves--;
                yCoordinate = (directions[directionIndex].equals("right")) ? yCoordinate+1 : yCoordinate-1;
            } else {
                for (int i=0; i<verticalMaxMoves; i++) {
                    finalList.add(matrix[yCoordinate][xCoordinate]);
                    yCoordinate = (directions[directionIndex].equals("down")) ? yCoordinate+1 : yCoordinate-1;
                    totalIterations--;
                }
                yCoordinate = (directions[directionIndex].equals("down")) ? yCoordinate-1 : yCoordinate+1;
                verticalMaxMoves--;
                xCoordinate = (directions[directionIndex].equals("down")) ? xCoordinate-1 : xCoordinate+1;
            }
            directionIndex = (directionIndex == 3) ? 0 : directionIndex+1;
        }

        return finalList;
    }
}
